### PR TITLE
add octave layer

### PR DESCRIPTION
**Pull request by izahn**
This PR adds an octave layer to spacemacs. It is fairly minimal at the moment, adding only activation of `octave-mode` for `*.m` files and setting up basic spacemacs REPL interaction keys. A README.org is included.

### CODE REVIEW

**Comment by robbyoconnor** [comment]
:+1:

**Comment by StreakyCobra** [comment]
Nice layer :smile:

Some information are missing though, like the dependencies and eventually a note about PATH if needed, have a look [here](https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Blang/haskell#dependencies) for an example.

AFAIK octave and matlab are pretty compatible, so it may be interesting to group them, as there is already a basic support for matlab [here](https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Blang/extra-langs).

**Comment by izahn** [comment]
Thanks @StreakyCobra, I've added some documentation about the dependencies. On the emacs side only `octave-mode` is required, and this is part of GNU Emacs, so there are no external elisp dependencies. I have clarified in the README that octave must be installed and in the PATH, as you suggested.

I'm not sure what you have in mind regarding integrating octave-mode and matlab-mode. To me they seem mutually exclusive, in the sense that both octave and matlab use `.m` file extensions, and you can only choose one major mode to be activated when such a file is visited. I'm open to suggestions of course!

**Comment by StreakyCobra** [comment]
Indeed, I was speaking about the `octave` dependency and the `$PATH` . Thanks for the update.

For octave and matlab… nevermind, I was just thinking out loud :-)

There was an update recently of the README's template for a better integration with the new documentation system. Can you also update it to the new template?

**Comment by izahn** [comment]
@StreakyCobra I've updated the README.org according to https://github.com/syl20bnr/spacemacs/blob/develop/core/templates/layer-README.template, though I'm a bit confused about it. I don't think the link syntax illustrated there is valid org-mode syntax. I did the best I could to follow the conventions, please let me know if any further changes are needed.

**Comment by StreakyCobra** [comment]
@izahn You can have a look [here](https://github.com/syl20bnr/spacemacs/blob/develop/doc/DOCUMENTATION.org) and indeed links are broken. But it's because documentation is moving in a new format, [example here](http://boyanangelov.com/spacemacs/layers/shell/README.html)

**Comment by StreakyCobra** [comment]
@izahn You can have a look [here](https://github.com/syl20bnr/spacemacs/blob/develop/doc/DOCUMENTATION.org) and indeed links are broken. But it's because documentation is moving in a new format, [example here](http://boyanangelov.com/spacemacs/layers/shell/README.html)

File: layers/+lang/octave/README.org [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+(setq-default dotspacemacs-configuration-layers '(octave))
+#+END_SRC
+
+* Key bindings
+
+** Inferior REPL process
+Send code to inferior process with these commands:
+
+| Key     | Description                                                  |
+|---------+--------------------------------------------------------------|
```
Comment (StreakyCobra) [inline layers/+lang/octave/README.org]: Can you also add the `SPC` prefix in this array?

File: layers/+lang/octave/packages.el [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    (evil-leader/set-key-for-mode 'octave-mode
+      "msi" 'run-octave
+      ;; helpers
+      "mhd" 'octave-help
+      "mhi" 'octave-lookfor
+      ;; REPL
+      "msb" 'octave-send-buffer
+      "msf" 'octave-send-defun
+      "msl" 'octave-send-line
+      "mlr" 'octave-send-region
```
Comment (StreakyCobra) [inline layers/+lang/octave/packages.el]: This binding  should probably be `"msr"` ;-)

File: layers/+lang/octave/packages.el [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ))
+
+(defun octave/init-octave ()
+  (use-package octave
+    :mode ("\\.m\\'" . octave-mode)
+    :commands (run-octave))
+
+  (with-eval-after-load 'octave
+    (evil-leader/set-key-for-mode 'octave-mode
+      "msi" 'run-octave
```
Comment (StreakyCobra) [inline layers/+lang/octave/packages.el]: Can you sort this binding with the others `ms…` alphabetically?

File: layers/+lang/octave/README.org (hunk 1 above)
Comment (izahn) [inline layers/+lang/octave/README.org]: Done, thank you!

File: layers/+lang/octave/packages.el (hunk 2 above)
Comment (izahn) [inline layers/+lang/octave/packages.el]: Fixed, thank you!

File: layers/+lang/octave/packages.el (hunk 3 above)
Comment (izahn) [inline layers/+lang/octave/packages.el]: Fixed, thank you!

**Comment by izahn** [comment]
I've done some code and documentation cleanups suggested by @StreakyCobra . Thank you for the helpful review! I think this is ready to go now.

**Comment by StreakyCobra** [comment]
Thanks for the changes. Nearly done, I forgot one or two details last time :smile: :

- Add `:defer t` to avoid loading the package unnecessarly
- Remove the `:command` property (the package will be loaded anyway when entering the mode, so it's useless)
- Move what's inside `with-eval-after-load` to the `:config` property of `use-package`.

I hope it is last modifications.

**Comment by izahn** [comment]
Thanks again @StreakyCobra , moving everything to the `:config` property is much cleaner. I've done that, and added `:defer ` (though according to https://github.com/jwiegley/use-package#notes-about-lazy-loading this is not needed). I have not removed the `:command` property, because I often want to start an interactive REPL without having an associated script. With the current configuration all that takes is `SPC : run-o RET`. If I remove it I would need `SPC ff tmp.m RET SPC m s i`, which is not only more cumbersome but also creates and unwanted `tmp.m` buffer. So I think  having `:commands (run-octave)` is very useful, and I would like to keep it.

**Comment by StreakyCobra** [comment]
Then `:defer t` it's not needed. I knew that `:command` implied it, but I didn't knew for `:mode`. I'll go to sleep smarter tonight :smile:

I didn't thought about this usage, but it totally makes sense, keep the `:command`!

So, if you can just remove the `:defer`, align the first table in the `org` and squash your commits, it's ready! 

PS: Sorry for the extra work :relaxed:

**Comment by izahn** [comment]
@StreakyCobra I've removed `:defer t` and aligned the keybindings table in the README. No worries about the extra work, and **thank you** for your very helpful reviews and suggestions. I'm excited about getting this into Spacemacs!

**Comment by StreakyCobra** [comment]
:+1: Everything looks fine for me :relaxed: . PRs for new layers should pass under the :eyes: of @syl20bnr to be merged, but I'll label it as `Reviewed` to hopefully accelerate the process.

Thanks for your work!

**Comment by bsima** [comment]
:+1: This layer is awesome, and much needed

**Comment by syl20bnr** [comment]
This one and elfeed are the next PRs I'll look into :-)

**Comment by TheBB** [comment]
Please put Swift on that list, too.

**Comment by syl20bnr** [comment]
This layer is the next on the list :-)

**Comment by syl20bnr** [comment]
Thank you ! :+1: 
I replace `SPC m h d` by `SPC m h h` to stick to the convention (documentation for thing under point), not 100% sure about it though, feel free to correct back if needed. 
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**